src/Umbraco.Core/Scoping/ScopeProvider.cs

```diff
@@ -0,0 +1,352 @@
+
+            StaticScopeReference.Dispose();
+        }
+
+        /// <inheritdoc />
+        public ScopeContext Context
+        {
+            get { return AmbientContext; }
+        }
+
+#if DEBUG_SCOPES
+        // this code needs TLC
+        // the idea is to keep in a list all the scopes that have been created
+        // and remove them when they are disposed
+        // so we can track leaks
+
+        // helps identifying when non-httpContext scopes are created by logging the stack trace
+        //private void LogCallContextStack()
+        //{
+        //    var trace = Environment.StackTrace;
+        //    if (trace.IndexOf("ScheduledPublishing") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Scheduled Publishing");
+        //    else if (trace.IndexOf("TouchServerTask") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Server Registration");
+        //    else if (trace.IndexOf("LogScrubber") > 0)
+        //        LogHelper.Debug<ScopeProvider>("CallContext: Log Scrubber");
+        //    else
+        //        LogHelper.Debug<ScopeProvider>("CallContext: " + Environment.StackTrace);
+        //}
+
+        // helps identifying scope leaks by keeping track of all instances
+        public static readonly List<ScopeInfo> StaticScopeInfos = new List<ScopeInfo>();
+
+        public IEnumerable<ScopeInfo> ScopeInfos { get { return StaticScopeInfos; } }
+
+        //private static void Log(string message, UmbracoDatabase database)
+        //{
+        //    LogHelper.Debug<ScopeProvider>(message + " (" + (database == null ? "" : database.InstanceSid) + ").");
+        //}
+
+        public void Register(IScope scope)
+        {
+            if (StaticScopeInfos.Any(x => x.Scope == scope)) throw new Exception();
+            StaticScopeInfos.Add(new ScopeInfo(scope, Environment.StackTrace));
+        }
+
+        public static void RegisterContext(IScope scope, string context)
+        {
+            var info = StaticScopeInfos.FirstOrDefault(x => x.Scope == scope);
+            if (info == null)
+            {
+                if (context == null) return;
+                throw new Exception();
+            }
+            if (context == null) info.NullStack = Environment.StackTrace;
+            info.Context = context;
+        }
+
+        public void Disposed(IScope scope)
+        {
+            var info = StaticScopeInfos.FirstOrDefault(x => x.Scope == scope);
+            if (info != null)
+            {
+                // enable this by default
+                StaticScopeInfos.Remove(info);
+
+                // instead, enable this to keep *all* scopes
+                // beware, there can be a lot of scopes!
+                //info.Disposed = true;
+                //info.DisposedStack = Environment.StackTrace;
+            }
+        }
+#endif
+    }
+
+#if DEBUG_SCOPES
+    public class ScopeInfo
+    {
+        public ScopeInfo(IScope scope, string ctorStack)
```

Comment:
Needs equality members here, should have it's own unique ID GUID which can be used as the key in the KeyedCollection  mentioned above.